crates/sui-types/src/passkey_authenticator.rs

```diff
@@ -0,0 +1,309 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::crypto::PublicKey;
+use crate::crypto::Secp256r1SuiSignature;
+use crate::crypto::SuiSignatureInner;
+use crate::signature_verification::VerifiedDigestCache;
+use crate::{
+    base_types::{EpochId, SuiAddress},
+    crypto::{DefaultHash, Signature, SignatureScheme, SuiSignature},
+    digests::ZKLoginInputsDigest,
+    error::{SuiError, SuiResult},
+    signature::{AuthenticatorTrait, VerifyParams},
+};
+use fastcrypto::hash::{HashFunction, Sha256};
+use fastcrypto::rsa::{Base64UrlUnpadded, Encoding};
+use fastcrypto::secp256r1::{Secp256r1PublicKey, Secp256r1Signature};
+use fastcrypto::traits::VerifyingKey;
+use fastcrypto::{error::FastCryptoError, traits::ToFromBytes};
+use once_cell::sync::OnceCell;
+use passkey::types::webauthn::{ClientDataType, CollectedClientData};
+use schemars::JsonSchema;
+use serde::{Deserialize, Deserializer, Serialize};
+use shared_crypto::intent::Intent;
+use shared_crypto::intent::{IntentMessage, INTENT_PREFIX_LENGTH};
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::sync::Arc;
+
+#[cfg(test)]
+#[path = "unit_tests/passkey_authenticator_test.rs"]
+mod passkey_authenticator_test;
+
+/// An passkey authenticator with parsed fields. See field defition below. Can be initialized from [struct RawPasskeyAuthenticator].
+#[derive(Debug, Clone, JsonSchema)]
+pub struct PasskeyAuthenticator {
+    /// `authenticatorData` is a bytearray that encodes
+    /// [Authenticator Data](https://www.w3.org/TR/webauthn-2/#sctn-authenticator-data)
+    /// structure returned by the authenticator attestation
+    /// response as is.
+    authenticator_data: Vec<u8>,
+
+    /// `clientDataJSON` contains a JSON-compatible
+    /// UTF-8 encoded serialization of the client
+    /// data which is passed to the authenticator by
+    ///  the client during the authentication request
+    /// (see [CollectedClientData](https://www.w3.org/TR/webauthn-2/#dictdef-collectedclientdata))
+    client_data_json: Vec<u8>,
```

Comment:
If this is supposed to be UTF8 encoded, then lets change this to a `String` type?

Reply:
i went back and forth on this, vec is slightly more compact but we can do string so that it will be validated before submitting to sui

Reply:
> vec is slightly more compact

Whats more compact about Vec? Vec and String have the serialized form only difference is that String is utf8 validated on deserialize

Reply:
the tradeoff here is that we do need the bytes form of client data json when computing `authenticator_data || sha256(client_data_json)`. it will require the string to be converted back to bytes. is it worth doing this on client side and convert it back?

Reply:
>it will require the string to be converted back to bytes

there's no overhead for converting a string to bytes, String is essentially just a wrapper over a Vec<u8> and you can easily get it into a bytes form via https://doc.rust-lang.org/stable/std/string/struct.String.html#method.as_bytes

Reply:
i see! thanks for the reference - revised to use string to enforce utf8 upon deserialization